lib/android_apk/aapt2/dump_badging.rb

```diff
@@ -3,120 +3,171 @@
+              raise InvalidApkError, "AndroidManifest.xml seems to be invalid and not decode-able."
+            else
+              # *normally* failed. The output of aapt2 dump is helpful.
+              # ref: https://cs.android.com/android/platform/superproject/+/master:frameworks/base/tools/aapt/Command.cpp;l=860?q=%22dump%20failed%22%20aapt
+              raise Aapt2Error.new(message: "This apk file cannot be parsed using 'aapt2 dump badging --include-meta-data'", stdout: stdout, stderr: stderr)
+            end
           else
             # unexpectedly failed. This may happen due to the running environment.
-            raise UnacceptableApkError, "'aapt2 dump badging --include-meta-data' failed due to an unexpected error."
+            raise Aapt2Error.new(message: "'aapt2 dump badging --include-meta-data' failed due to an unexpected error.", stdout: stdout, stderr: stderr)
           end
         end
       end
 
       # @param apk_filepath [String] a path to apk_filepath
       def initialize(apk_filepath:)
         @apk_filepath = apk_filepath
-        @dump_results = self.class.dump_badging(apk_filepath: apk_filepath)&.scrub&.split("\n")
       end
 
       # Parse output of aapt2 command to Hash format
       #
       # @return [::AndroidApk::Aapt2::DumpBadging::Result]
       def parse
-        vars = {}
-        results = @dump_results.dup
-        results.each do |line|
-          key, value = parse_line(line)
-          next if key.nil?
+        return @parse if defined?(@parse)
 
-          if vars.key?(key)
-            reject_illegal_duplicated_key!(key)
+        raw_result_lines = self.class.dump_badging(apk_filepath: @apk_filepath).scrub.split("\n")
 
-            if vars[key].kind_of?(Hash) and value.kind_of?(Hash)
-              vars[key].merge(value)
-            else
-              vars[key] = [vars[key]] unless vars[key].kind_of?(Array)
-              if value.kind_of?(Array)
-                vars[key].concat(value)
-              else
+        vars = {
+          "labels" => {},
+          "icons" => {},
+          "meta-data" => []
+        }
+
+        raw_result_lines.each do |line|
+          key, value = parse_line(line)
+
+          if !(m = key.match(/\Aapplication-icon-(\d+)\z/)).nil?
+            vars["icons"][m[1].to_i] = value unless value.nil?
+          elsif !(m = key.match(/\Aapplication-label-(\S+)\z/)).nil?
+            vars["labels"][m[1]] = value unless value.nil?
+          else
+            # noinspection RubyCaseWithoutElseBlockInspection
+            case key
+            when *BOOLEAN_ELEMENT_TAG_NAMES
+              vars[key] = true
+            when *MULTIPLE_ELEMENTS_TAG_NAMES
+              if value.kind_of?(Hash)
                 vars[key].push(value)
+              else
+                vars[key].push(*value)
               end
+            when *SINGLE_VALUE_ELEMENT_TAG_NAMES
+              reject_illegal_duplicated_key!(key) if vars.key?(key)
+
+              vars[key] = value
```

Comment:
Became simple because some unused processing like Hash merging have been removed and value transformation have been inlined.